### ISSUE TITLE
Re-run only updated `\py{}` blocks

### ISSUE THREAD

**Issue by TheChymera** [gpoore/pythontex#150]
Hi, I wanted to ask if there is any way to devolve re-running to the individual code blocks. I am writing increasingly large (e.g. a [recent article](https://www.biorxiv.org/content/10.1101/619650v1)) documents with PythonTeX, and I am noticing that it doesn't scale that well.

**Comment by gpoore** [comment]
There isn't a way to do that without caching the complete state for each code block, and then trying to track dependencies between code blocks. That's a difficult problem to solve.

If the scaling issue is due to creating plots, then I would suggest trying to put each plot, or groups of related plots, into separate sessions. So you would use something like `\py[plot1]{<code>}`. Sessions can run in parallel and are completely independent. If the scaling issue is due to calculations instead, then I would recommend saving the results of the calculations (something as simple as JSON might work, depending on what you're doing), and then loading the results and plotting them in a separate session.

**Comment by TheChymera** [comment]
Ah, I had no idea about the sessions part, it turns out everything is in one session, which might explain why it takes me half an hour to compile my document... Is there any reason why code blocks are not given their own session? is there any way a \py{<code>} could depend on another?

**Comment by gpoore** [comment]
A common use for `\py` is to get a string representation of a variable, or evaluate an expression inline. For that, you might want a preceding `pycode` environment to do some setup or calculations, so it's nice to have everything in a single session by default.

It's possible for one `\py` to depend on another if it evaluates an expression that has side effects, like changing a global variable or a class instance attribute.

If you want to always use each `\py` in its own session, you could create a new command. Maybe something like this (untested):
```latex
\newcounter{pysessioncounter}
\newcommand{\sessionpy}{%
  \edef\sessionpysession{session\arabic{pysessioncounter}}%
  \stepcounter{pysessioncounter}%
  \expandafter\py\expandafter[\sessionpysession]}
```

**Comment by obtitus** [comment]
I have written my entire Master and PhD thesis using PythonTeX and found that it does scale quite well. Some tips

For larger documents I would also recommend to split sessions on section, parts and/or chapters, by using the
```latex
\restartpythontexsession{\arabic{section}}
```
One big drawback is that when you re-arrange sections PythonTeX will re-run a lot of code.

If you have some common imports, functions and constants that you want to share globally for the entire document, use the pythontexcustomcode. Example of importing a file called pythonPreamble.py in a src directory:

```latex
\begin{pythontexcustomcode}[begin]{pylab}
sys.path.insert(0, "src/") 
pytex.add_dependencies('src/pythonPreamble.py')
\begin{pythontexcustomcode}[end]{pylab}
```
This will put this code snippet at the start of each pylab session.

For heavy calculations you can also save quite some time with the [joblib.Memory](https://joblib.readthedocs.io/en/latest/generated/joblib.Memory.html) python module. Example:

```python
from joblib import Memory
cachedir = 'tmp'
memory = Memory(cachedir=cachedir)

@memory.cache
def f_peak_cached(filename):
    m = Measurement(filename=filename)
    m = m.cut_start()
    f_peak = m.plot_psd(ax=None, f_max=4.9e9)
    return f_peak
```
This will ensure that f_peak_cached is only called once for every filename (does not detect if the file has changed!)

**Comment by TheChymera** [comment]
@obtitus so we meet again :) I have also written my MSc. thesis with it, and currently doing the same for my PhD. I'm not sure I understand your joblib example, though.

**Comment by TheChymera** [comment]
@gpoore 's proposed `\sessionpy` solution turns out to provide everything I need at the moment. Thank you very much!

**Comment by TheChymera** [comment]
@gpoore 

Hi again, I am noticing that the proposed solution:

```
\newcounter{pysessioncounter}
\newcommand{\sessionpy}{%
  \edef\sessionpysession{session\arabic{pysessioncounter}}%
  \stepcounter{pysessioncounter}%
  \expandafter\py\expandafter[\sessionpysession]}
```

Fails with:

```
! LaTeX Error: Command \c@pysessioncounter already defined.
               Or name \end... illegal, see p.192 of the manual.
```

I tried renaming the counter to `pysinglesessioncounter`, but that fails as well. Any idea what I could try?

**Comment by gpoore** [comment]
@TheChymera You probably want to do a search for `\newcounter{pysessioncounter}`. It's probably defined at multiple locations in the document.

**Comment by TheChymera** [comment]
@gpoore I thought the same thing, which is why I tried to use the name `pysinglesessioncounter`, and ended up getting the exact same error.

Querying the text ([this](https://github.com/TheChymera/RepSeP) being the repo) indicates that there's really just this one definition:

```shell
chymera@silenthost ~/src/RepSeP $ ag pysessioncounter
common_header.tex
29:\newcounter{pysessioncounter}
31:          \edef\sessionpysession{session\arabic{pysessioncounter}}%
32:            \stepcounter{pysessioncounter}%
```

**Comment by gpoore** [comment]
@TheChymera There has to be a double definition somehow to get that error, unless there was an earlier error in the log that did something really strange. Are there any `\input` or packages involving files in other directories that could cause this? Could `common_header` be included twice anywhere somehow? You might also try deleting the temp files from running latex.

**Comment by TheChymera** [comment]
@gpoore sadly nothing of the sort, as I said, it also keeps happening if I rename the counter... Are you sure it's the duplication? The error message also says `Or name \end... illegal`...

Would you be willing to try and reproduce this? It's actually a rather small pythontex-based example, just that it requires this additional file: https://github.com/TheChymera/repsep_utils/ since I decided to centrally distribute it, rather than include it in every package. If you want I'd be perfectly happy to submit a PR with this to PythonTeX.
